Mathlib/Data/Finsupp/SMulWithZero.lean

```diff
@@ -18,9 +18,17 @@ This file defines the pointwise scalar multiplication on `Finsupp`, assuming it
 
 ## Implementation notes
 
+This file is intermediate between `Finsupp.Defs` and `Finsupp.Module` in that it covers scalar
+multiplication but does not rely on the definition of `Module`. Scalar multiplication is needed to
+supply the `nsmul` (and `zsmul`) fields of (semi)ring structures which are fundamental for e.g.
+`Polynomial`, so we want to keep the imports requied for the `Finsupp.smulZeroClass` instance
```

Comment:
I'm sorry, not sure how to respond except for saying that my intuition would group `Module R R[X]` with `Algebra R R[X]` (except for the commutativity in `R`). But of course arguing about intuition would get us nowhere. What kind of standards should we use to decide this?

Reply:
Agreed that `Algebra R R[X]` is just as fundamental here (incidentally, I find it rather annoying that the `Algebra R R[X]` instead is segregated away in `Mathlib.Algebra.Polynomial.AlgebraMap`, which makes it easy to forget to import).

I also agree that subjective judgements about intuition and things being "fundamental" isn't a great way to handle library design!

Is this PR part of some larger re-structuring goal? If so, it would be great to have a tracking issue summaring the main aims; at the moment, I'm getting distracted by the details because I don't know what the big picture is!

Reply:
The ultimate goal is to have a large chunk of definitions in Mathlib that can be imported without referencing any more theory than necessary. (Of course, the meaning of "definition" as opposed to "theory" is rather vague. And I suppose that's exactly the issue here: is `Module R R[X]` a definition or "theory"?) I have discussed the project mostly with Kim, but I'll make a tracking issue for that.

Reply:
In this specific case, the motivation is that the import graph up to `Algebra.Polynomial.Eval.Defs` has a lot of grey nodes in `LinearAlgebra` and `Algebra.Module`. So avoiding `Module` when defining `Polynomial` should clear up that dependency a lot.

Reply:
Following up on your comment about `Algebra R R[X]` being fundamental and currently hidden away in `Polynomial.AlgebraMap`, does PR #19281 moving it to a dedicated file help?

Reply:
No I think that makes it worse, my preference would be for the algebra to be defined in the same file as all the other scalar action structure (and preferably also as the ring structure). Otherwise you can end up with a goal about scalar multiplication, and be confused about why lemmas like `Algebra.smul_def` aren't applying, and it's hard to realize the issue is a missing import.

So I guess generally the rule I'd like to see is "if you import `Basic` for an algebraic type, then you get all the "basic" scalar action and algebra instances". This obviously leaves room for `Defs` to be missing some instances, on the premise that users shouldn't be importing `Defs`. `Finsupp` is arguably not primarily algebraic, so I guess there's an argument that the basic file doesn't need to define even the `AddMonoid` structure.

Reply:
Hmm. I mean, we could literally rename `Polynomial/Algebra.lean` to `Basic.lean` (and `MonoidAlgebra/Algebra.lean` to `Basic.lean` if we also rename the existing `Basic.lean` to `Lemmas.lean`, which I'm not unopposed to doing) and we'd basically have what you are asking for. Would that help? :)

(Unfortunately `Polynomial` still has a few instances like `IsDomain` quite far away from basic definitions since they go through `degree`/`eval`...)

Reply:
I'm not sure what the status of this PR is now. If the issue is with this comment, then I'd be happy to change it, but if it is with the splitting philosophy then we'd better open a public Zulip discussion.

Reply:
@eric-wieser, would you have some time to come back to this?

Reply:
@Vierkantor, how about we do some renaming so that importing `Basic` ensures all the instances available, and then otherwise proceed as is?

Reply:
I have reworked #19281 so that this happens for the PRs depending on this one (namely putting the algebra structure for `Polynomial` and `MonoidAlgebra` in their respective `Basic.lean`). For `Finsupp` the current state of its `Basic` file is still too messy to make that practical.